draft/keymanagement.md

```diff
@@ -0,0 +1,129 @@
+# New acpeerstates table
+
+A proposal to replace `acpeerstates` table,
+which has columns
+- `id`
+- `addr`
+- `prefer_encrypted`
+- `last_seen`
+- `last_seen_autocrypt`
+- `public_key`
+- `public_key_fingrprint`
+- `gossip_timestamp`
+- `gossip_key`
+- `gossip_key_fingerprint`
+- `verified_key`
+- `verified_key_fingerprint`
+- `verifier`
+- `secondary_verified_key`
+- `secondary_verified_key_fingerprint`
+- `secondary_verifier`
+with a new `public_keys` table with columns
+- `id`
+- `addr` - address of the contact which is supposed to have the private key
+- `public_key` - a blob with the OpenPGP public key
+- `public_key_fingerprint` - public key fingerprint
+- `introduced_by` - address of the contact if received directly in `Autocrypt` header, address of `Autocrypt-Gossip` sender otherwise
+- `is_verified` - boolean flag indicating if the key was received in a verified group or not
+- `timestamp` - timestamp of the most recent row update 
+
+The table has consraints `UNIQUE(addr, introduced_by)`.
+Note that there may be multiple rows with the same fingerprint,
+e.g. if multiple peers gossiped the same key for a contact they all are introducers of the key.
+
+`prefer_encrypted` is moved to the `contacts` table.
+
+
+## Using the table to select the keys
+
+### Sending a message to 1:1 chat
+
+When sending a message to the contact in a 1:1 chat,
+encrypt to the key where `introduced_by` equals `addr`:
+```
+SELECT *
+FROM public_keys
+WHERE addr=? AND introduced_by=addr
+```
+This row is guaranteed to be unique if it exists.
+
+If direct Autocrypt key does not exist,
+use the most recent one according to the timestamp key gossiped for this contact:
+```
+SELECT *
+FROM public_keys
+WHERE addr=?
+ORDER BY timestamp DESC LIMIT 1
+```
+
+If the table contains a row with selected key and `is_verified` flag set,
+display a green checkmark and "Introduced by <introduced_by>" in the 1:1 chat/contact profile.
+Note that row with verification may be older than the most recent gossip.
+
+### Sending a message to unprotected group chat
+
+When sending a message to the contact in a non-verified group chat,
+encrypt to the same key as used in 1:1 chat
+and (NEW!) one or more most recent gossiped keys introduced by chat members:
+```
+SELECT public_key
+FROM public_keys
+WHERE addr=? AND introduced_by IN (<chat-member-list>)
+ORDER BY timestamp DESC
+```
+It does not matter if the gossiped key has `is_verified` flag.
+
+### Sending a message to protected group chat
+
+When sending a message in a protected group chat,
+construct a list of candidate keys for each contact
+the same way as in an unprotected chat,
+but then filter out the keys which do not have `is_verified` flag.
```

Comment:
If we also encrypt to old keys, we make it impossible to "revoke" a compromised key by just re-installing delta chat.

Consider Alice, who has her unencrypted phone confiscated with all her green group chats; she changes the passphrase to her email account and re-installs Delta Chat to create a new keypair. Then she re-authenticates with her friends to be able to read the messages in the green groups again. Everyone would still encrypt messages to her compromised key, so when the police requests her mail server for her encrypted emails, they would still be able to read them.

Not a super likely case, but only encrypting to these two recent keys might avoid that:

```
SELECT public_key
FROM public_keys
WHERE addr=? AND is_verified=True AND introduced_by IN (<chat-member-list>)
ORDER BY timestamp DESC LIMIT 1
```

and

```
SELECT public_key
FROM public_keys
WHERE addr=? AND is_verified=True AND introduced_by=addr
ORDER BY timestamp DESC
```

This should correspond with the primary_verified and secondary_verified keyslots from #4898.

Reply:
In addition to what @missytake said, if someone reinstalls DC multiple times, then there may be lots of keys in the table (think about a group with 40 members - I can imagine the list of keys we need to encrypt to grow to 100, if we have a bit more than 2 keys per recipient on average).

What about this:
- clear all indirect unverified keys after we get an Autocrypt message directly from a contact, and 
- clear all indirect keys after a securejoin

We know that this is the key the contact is using now, and it is unlikely that they are using another key at the same time.

Reply:
>     * clear all indirect keys after a securejoin

i guess you mean on the side of alice and bob.  Other group members can not clear gossip keys until they saw a direct message from Bob.  Until then they will gossip the wrong bob key and even Alice will memorize it shortly after the scan.   Note that i suggested elsewhere we consider a securejoin scan to result in a Bob broadcast of "i joined" system message which will clear everyone's gossiped Bob keys from the group member's  tables, making the alice-introduction the one and only verified key that gets gossiped.

Reply:
Maybe just use this query by @missytake:
```
SELECT public_key
FROM public_keys
WHERE addr=? AND is_verified=True AND introduced_by IN (<chat-member-list>)
ORDER BY timestamp DESC LIMIT 1
```
This way the new Bob's key gossiped by Alice in `vg-member-added` will be chosen. (But if there's a newer unverified key introduced by Bob himself, just not encrypt to Bob in this group at all because the older verified key may have been seized.)

Afaiu with "i joined" we still can only clear the older Bob's key introduced by Alice, because if Alice steals Bob's email account, that shouldn't affect other protected groups where Alice isn't a member.